moved transaction_records_by_id methods into it

fluidvanadium opened the pull request:
This is a step toward implementing input source.

A number of unit tests are incoming

fluidvanadium committed:
moved transaction_records_by_id methods into it

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

zancas commented on zingolib/src/wallet/transactions/recording.rs:
```diff
@@ -24,85 +24,18 @@ use crate::{
-        for txid in &txids_to_remove {
-            self.transaction_records_by_id.remove(txid);
-        }
+    pub fn invalidate_txids(&mut self, txids_to_remove: Vec<TxId>) {
```
I find this pattern of the external type exposing the exact same interface as the internal wrap smelly...    it sniffs funny.

fluidvanadium replied:
actually, its not exact. the outer type deals with witness trees too.

fluidvanadium replied:
otherwise i would just skip this layer

Oscar-Pepper replied:
I dont understand this reply. why can't we call `invalidate_txids` directly on `transaction_records_by_id` instead of adding this empty wrapper?

fluidvanadium replied:
oops, there was an empty wrapper there! fixed!

zancas commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -21,8 +31,94 @@ impl std::ops::DerefMut for TransactionRecordsById {
     }
+
+    // modify methods
+
```
Please add a unit-test to this method, or reference an open PR that does so?  Same for each new method.

fluidvanadium replied:
#945 

however, this method is not new. just reorganized

zancas commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -21,8 +31,94 @@ impl std::ops::DerefMut for TransactionRecordsById {
+            .filter_map(|transaction_metadata| {
+                if transaction_metadata
+                    .status
+                    .is_confirmed_after_or_at(&reorg_height)
```
Should we have a simple wrapper that's `after_or_at` that combines broadcast and confirmed..  or is better to keep them more separate and explicit at this layer of abstraction?

zancas commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -21,8 +31,94 @@ impl std::ops::DerefMut for TransactionRecordsById {
+            self.remove(txid);
+        }
+
+        // invalidate (roll back) any transparent spends in each invalidated tx
```
Why not make invalidation of transparent utxos a method, that's analogous to the other pool-specific invalidation methods?

fluidvanadium replied:
okay, we will come back over this with another layer of polish

fluidvanadium committed:
Addressed review comments:

Named functions and parameters for clarity
Helperized
new constructor
Added doc comments

fluidvanadium committed:
clippy auto

fluidvanadium committed:
clippy p2

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
     }
 }
+
+/// This block implements constructors.
```
my opinion is we dont need to comment impl blocks as its explicit from the function declarations
(recommended, not required)

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+
+/// This block implements constructors.
+impl Default for TransactionRecordsById {
+    fn default() -> Self {
```
doc comment
```suggestion
    /// Default constructor
    fn default() -> Self {
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
         TransactionRecordsById(map)
     }
 }
+/// This block implements methods to modify the map.
```
see above comment on impl comments

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+}
+
 impl TransactionRecordsById {
+    // New empty Map
```
doc comment
```suggestion
    /// Constructs a new TransactionRecordsById with an empty map.
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
 }
+/// This block implements methods to modify the map.
+impl TransactionRecordsById {
+    /// All information after a certain height is invalidated during a reorg.
```
add function description above these extra details. also when talking about height I think we should use `above` or `higher` instead of `after`
```suggestion
    /// Invalidates all transactions from a given height including the block with block height `reorg_height`
    ///
    /// All information above a certain height is invalidated during a reorg.
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+/// This block implements methods to modify the map.
+impl TransactionRecordsById {
+    /// All information after a certain height is invalidated during a reorg.
+    pub fn invalidate_all_transactions_after_or_at_height(&mut self, reorg_height: BlockHeight) {
```
does this function need such a cumbersome name?
```suggestion
    pub fn truncate_transactions(&mut self, reorg_height: BlockHeight) {
```

fluidvanadium replied:
yes

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+            .collect::<Vec<_>>();
+        self.invalidate_txids(txids_to_remove);
+    }
+    /// this function invalidiates a vec of txids by removing them and then all references to them.
```
not required, I think looks more polished to remove `this function` etc.
```suggestion
    /// Invalidiates a vec of txids by removing them and then all references to them.
```

Oscar-Pepper replied:
UPDATE
```suggestion
    /// Invalidiates transactions in the map by txid
    ///
    /// Takes a vec of txids corresponding to the transactions for removal
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+    pub fn new() -> Self {
+        TransactionRecordsById(HashMap::new())
+    }
     // Associated function to create a TransactionRecordMap from a HashMap
```
missing doc comment
```suggestion
    /// Constructs a TransactionRecordMap from a HashMap
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+        self.invalidate_txids(txids_to_remove);
+    }
+    /// this function invalidiates a vec of txids by removing them and then all references to them.
+    pub(crate) fn invalidate_txids(&mut self, txids_to_remove: Vec<TxId>) {
```
this invalidates a transaction but takes a txid right?
```suggestion
    pub(crate) fn invalidate_transactions(&mut self, txids_to_remove: Vec<TxId>) {
```

fluidvanadium replied:
true

fluidvanadium replied:
addressed

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+        // roll back any orchard spends in each invalidated tx
+        self.invalidate_txid_specific_domain_spends::<OrchardDomain>(&txids_to_remove);
+    }
+    /// any transparent note that is listed as being spent in one of these transactions will be marked as unspent.
```
suggested comment change
```suggestion
    /// Reverts any spent transparent notes in the given transactions to unspent.
    ///
    /// This is required in the case that the note was spent in an invalidated transaction.
    /// Takes a slice of txids corresponding to the invalidated transactions.
```

Oscar-Pepper commented on zingolib/src/wallet/transaction_records_by_id.rs:
```diff
@@ -20,9 +30,108 @@ impl std::ops::DerefMut for TransactionRecordsById {
+                })
+        });
+    }
+    /// any shielded note of the specified domain that is listed as being spent in one of these transactions will be marked as unspent.
```
suggested comment change 
```suggestion
    /// Reverts any spent shielded notes in the given transactions to unspent.
    ///
    /// This is required in the case that the note was spent in an invalidated transaction.
    /// Takes a slice of txids corresponding to the invalidated transactions.
```

Oscar-Pepper commented on zingolib/src/wallet/transactions/recording.rs:
```diff
@@ -24,85 +24,18 @@ use crate::{
+            .invalidate_txids(txids_to_remove)
     }
 
     // During reorgs, we need to remove all txns at a given height, and all spends that refer to any removed txns.
```
missing doc comment
```suggestion
    /// During reorgs, we need to remove all txns at a given height, and all spends that refer to any removed txns.
```

Oscar-Pepper reviewed (dismissed):
this is great, very concise and I know exactly what im reveiewing now.

I have left a bunch of small requests / suggestions.

fluidvanadium committed:
Update zingolib/src/wallet/transactions/recording.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Update zingolib/src/wallet/transaction_records_by_id.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Update zingolib/src/wallet/transaction_records_by_id.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Update zingolib/src/wallet/transaction_records_by_id.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Update zingolib/src/wallet/transaction_records_by_id.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Update zingolib/src/wallet/transaction_records_by_id.rs

Co-authored-by: oscar-pepper <109323234+Oscar-Pepper@users.noreply.github.com>
Signed-off-by: fluidvanadium <fluidvanadium@pm.me>

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

fluidvanadium committed:
clarified some function names. added some doc-comments. removed an accessory function

fluidvanadium committed:
improve doc comments

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

fluidvanadium committed:
even clearer doc-comment

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

fluidvanadium commented:
only changes private structs, so no changelog is needed

fluidvanadium committed:
Merge remote-tracking branch 'labs/dev' into move_methods_to_transaction_records_by_id

Oscar-Pepper reviewed (approved):
great!
